### PR TITLE
Migrate backup code from the k8s repository

### PR DESCRIPTION
**Pull request by shayancanonical**
[DPE-1194](https://warthogs.atlassian.net/browse/DPE-1194)

## Issue
1. We have implemented backups in the k8s repository, but have to migrate it to this repository
2. The core of the backup and restore logic is very similar, and we dont want to duplicate it in both repositories 

## Solution
1. Migrate it to this repo
2. Move the core of the backup logic into the shared charm lib (so it can be accessed by both repos)

[DPE-1194]: https://warthogs.atlassian.net/browse/DPE-1194?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by paulomach — COMMENTED** [review]
@shayancanonical , is this missing some push? Shouldn't `s3_helpers.py` and `backup.py` be moved to a lib?

**Comment by shayancanonical** [comment]
@paulomach I hadnt considered it, but have moved s3_helpers.py and backups.py to libraries in [18ab1dc](https://github.com/canonical/mysql-operator/pull/140/commits/18ab1dc9c5d87c46e26e4d97eb57eeaecd60771b)

File: src/constants.py [hunk 1]
```diff
@@ -27,5 +27,10 @@
 CHARMED_MYSQL_COMMON_DIRECTORY = "/var/snap/charmed-mysql/common"
 MYSQLD_SOCK_FILE = f"{CHARMED_MYSQL_COMMON_DIRECTORY}/mysql/mysqld.sock"
 MYSQLD_CONFIG_DIRECTORY = f"{CHARMED_MYSQL_COMMON_DIRECTORY}/mysql/conf.d"
+MYSQLD_DEFAULTS_CONFIG_FILE = f"{CHARMED_MYSQL_COMMON_DIRECTORY}/mysql/mysql.cnf"
```
Comment (delgod) [inline src/constants.py:30]: hm, can we follow snap structure according to the spec https://docs.google.com/document/d/1MyT4Wt_-vgPwhwEjC-oeHRdLYrZWaQzFGTwclT7hD1U/edit

**Review by delgod — CHANGES_REQUESTED** [review]

File: src/constants.py (hunk 1 above)
Comment (shayancanonical) [reply src/constants.py:30]: @jardon mentioned that we will be working on restructuring the charmed-mysql snap next pulse. captured the effort in [dpe-1459](https://warthogs.atlassian.net/browse/DPE-1459)

File: src/mysql_vm_helpers.py [hunk 2]
```diff
@@ -134,6 +145,52 @@ def install_and_configure_mysql_dependencies() -> None:
...
+
+        if innodb_buffer_pool_chunk_size:
+            content.append(f"innodb_buffer_pool_chunk_size = {innodb_buffer_pool_chunk_size}")
+        content.append("")
+
+        # create the mysqld config directory if it does not exist
+        logger.debug("Copying custom mysqld config")
+        pathlib.Path(MYSQLD_CONFIG_DIRECTORY).mkdir(mode=0o755, parents=True, exist_ok=True)
+
+        with open(f"{MYSQLD_CONFIG_DIRECTORY}/z-custom-mysqld.cnf", "w+") as config_file:
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/mysql_vm_helpers.py]: addressed in c2c501fc94feb3036aef0467221d4c965a30cb5e

File: src/mysql_vm_helpers.py [hunk 3]
```diff
@@ -197,6 +254,83 @@ def wait_until_mysql_connection(self) -> None:
...
+    def execute_backup_commands(
+        self,
+        s3_bucket: str,
+        s3_directory: str,
+        s3_access_key: str,
+        s3_secret_key: str,
+        s3_endpoint: str,
+    ) -> Tuple[str, str]:
+        """Executes commands to create a backup."""
+        return super(MySQL, self).execute_backup_commands(
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/mysql_vm_helpers.py]: addressed in c2c501fc94feb3036aef0467221d4c965a30cb5e

File: src/mysql_vm_helpers.py [hunk 4]
```diff
@@ -197,6 +254,83 @@ def wait_until_mysql_connection(self) -> None:
...
+            MYSQLD_SOCK_FILE,
+            f"{CHARMED_MYSQL_COMMON_DIRECTORY}/mysql",
+            MYSQLD_DEFAULTS_CONFIG_FILE,
+            user=MYSQL_SYSTEM_USER,
+            group=MYSQL_SYSTEM_USER,
+        )
+
+    def delete_temp_backup_directory(self) -> None:
+        """Delete the temp backup directory."""
+        super(MySQL, self).delete_temp_backup_directory(
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/mysql_vm_helpers.py]: addressed in c2c501fc94feb3036aef0467221d4c965a30cb5e

File: src/mysql_vm_helpers.py [hunk 5]
```diff
@@ -197,6 +254,83 @@ def wait_until_mysql_connection(self) -> None:
...
+
+            stdout = subprocess.check_output(
+                commands,
+                user=user,
+                group=group,
+                env=env,
+                stderr=subprocess.PIPE,
+                encoding="utf-8",
+            ).strip()
+            return (stdout, "")
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/mysql_vm_helpers.py]: addressed in c2c501fc94feb3036aef0467221d4c965a30cb5e

**Review by taurus-forever — APPROVED** [review]

File: src/mysql_vm_helpers.py [hunk 6]
```diff
@@ -197,6 +254,83 @@ def wait_until_mysql_connection(self) -> None:
...
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/mysql_vm_helpers.py:330]: the `_execute_commands` is an abstract method (to be implemented differently in vm and k8s). in k8s, we run `pebble exec` which returns both stdout and stderr as different streams. with subprocess, we have to (and i will double check this) pass in a file-like object if we want to capture stderr differently

[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/mysql_vm_helpers.py]: good catch, i thought i had passed in `stderr=subprocess.STDOUT` but had actually passed in `stderr=subprocess.PIPE`. also, the usage of run is far better, which i have changed to in f0df0128a32ac5035d1ebaba50ca33f5ef7b20f7

File: lib/charms/mysql/v0/mysql.py [hunk 7]
```diff
@@ -1252,6 +1271,186 @@ def offline_mode_and_hidden_instance_exists(self) -> bool:
...
+        try:
+            logger.info("Retrieving the total memory of the server")
+
+            """Below is an example output of `free --bytes`:
+               total        used        free      shared  buff/cache   available
+Mem:     16484458496 11890454528   265670656  2906722304  4328333312  1321193472
+Swap:     1027600384  1027600384           0
+            """
+            # need to use sh -c to be able to use pipes
+            get_total_memory_command = "free --bytes | sed -n '2p' | awk '{print $2}'".split()
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply lib/charms/mysql/v0/mysql.py]: Addressed in f0df0128a32ac5035d1ebaba50ca33f5ef7b20f7

File: src/mysql_vm_helpers.py [hunk 8]
```diff
@@ -197,6 +254,84 @@ def wait_until_mysql_connection(self) -> None:
...
+        """Execute commands on the server where mysql is running.
+
+        Args:
+            commands: a list containing the commands to execute
+            bash: whether to run the commands with bash
+            user: the user with which to execute the commands
+            group: the group with which to execute the commands
+            env: the environment variables to execute the commands with
+
+        Returns: tuple of (stdout + stderr, "")
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/mysql_vm_helpers.py]: updated in fe7e667ba3ee2987ed53d152e81768c5eaba6376

File: lib/charms/mysql/v0/s3_helpers.py [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+    s3_client = boto3.client(
+        "s3",
+        aws_access_key_id=s3_parameters["access-key"],
+        aws_secret_access_key=s3_parameters["secret-key"],
+        endpoint_url=s3_parameters["endpoint"],
+        region_name=s3_parameters["region"],
+    )
+
+    try:
+        response = s3_client.get_object(Bucket=s3_parameters["bucket"], Key=path)
```
Comment (delgod) [inline lib/charms/mysql/v0/s3_helpers.py:148]: if we check that file exists, why we need to download it all?
```suggestion
        response = s3_client.get_object(Bucket=s3_parameters["bucket"], Key=path, Range="0-1")
```

can be handled in separate PR

**Review by delgod — APPROVED** [review]

**Review by delgod — APPROVED** [review]

**Comment by shayancanonical** [comment]
Closing, as the code in this branch was merged as part of [PR 143](https://github.com/canonical/mysql-operator/pull/143)

**Pull request closed without merging**